x/gamm/twap/types/utils.go

```diff
@@ -0,0 +1,82 @@
+package types
+
+import (
+	fmt "fmt"
+	"sort"
+	time "time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/osmosis-labs/osmosis/v10/osmoutils"
+)
+
+func NewTwapRecord(k AmmInterface, ctx sdk.Context, poolId uint64, denom0 string, denom1 string) TwapRecord {
+	if !(denom0 > denom1) {
+		panic(fmt.Sprintf("precondition denom0 > denom1 not satisfied. denom0 %s | denom1 %s", denom0, denom1))
+	}
+	sp0 := MustGetSpotPrice(k, ctx, poolId, denom0, denom1)
+	sp1 := MustGetSpotPrice(k, ctx, poolId, denom1, denom0)
+	return TwapRecord{
+		PoolId:                      poolId,
+		Asset0Denom:                 denom0,
+		Asset1Denom:                 denom1,
+		Height:                      ctx.BlockHeight(),
+		Time:                        ctx.BlockTime(),
+		P0LastSpotPrice:             sp0,
+		P1LastSpotPrice:             sp1,
+		P0ArithmeticTwapAccumulator: sdk.ZeroDec(),
+		P1ArithmeticTwapAccumulator: sdk.ZeroDec(),
+	}
+}
+
+// mustGetSpotPrice returns the spot price for the given pool id, and denom0 in terms of denom1.
+// Panics if the pool state is misconfigured, which will halt any tx that interacts with this.
+func MustGetSpotPrice(k AmmInterface, ctx sdk.Context, poolId uint64, baseAssetDenom string, quoteAssetDenom string) sdk.Dec {
+	sp, err := k.CalculateSpotPrice(ctx, poolId, baseAssetDenom, quoteAssetDenom)
+	if err != nil {
+		panic(err)
+	}
+	return sp
+}
+
+// GetAllUniqueDenomPairs returns all unique pairs of denoms, where for every pair
+// (X, Y), X >= Y.
+// The pair (X,Y) should only appear once in the list
+//
+// NOTE: Sorts the input denoms slice.
+func GetAllUniqueDenomPairs(denoms []string) ([]string, []string) {
```

Comment:
consider changing these to named return values

Reply:
I think named return values are dangerous in large functions because it makes it riskier to return something unintended due to accidentally assigning something wrong to a named return

No strong opinions here though

Reply:
Agreed on not wanting to do a named return here